File: 600/670 - Maximum Swap.py
class Solution:
    def maximumSwap(self, num):
        """
        :type num: int
        :rtype: int
        """
        s = list(str(num))
        for i in range(len(s)):
            if i < len(s) - 1 and s[i] < s[i + 1]:
                break
        if i == len(s) - 1:
            return num
        m = s[i + 1]
        mi = i + 1
        for j, c in enumerate(s[i + 1:]):
            if c >= m:
                m = c
                mi = i + 1 + j

        for i in range(i + 1):
            if s[i] < m:
                s[mi] = s[i]
                s[i] = m
                break
        return int(''.join(s))

File: csharp/670.maximum-swap.cs
/*
 * @lc app=leetcode id=670 lang=csharp
 *
 * [670] Maximum Swap
 */

// @lc code=start
public class Solution {
    public int MaximumSwap(int num) {
        int[] last = new int[10];
        var s = num.ToString().ToCharArray();
        int n = s.Length;
        for (int i = 0; i < n; ++i) {
            last[(int) s[i] - (int) '0'] = i;
        }
        for (int i = 0; i < n; ++i) {
            int curr = (int) s[i] - (int) '0';
            for (int j = 9; j > curr; --j) {
                if (last[j] <= i) {
                    continue;
                }
                s[last[j]] = (char) (curr + (int) '0');
                s[i] = (char) (j + (int) '0');
                return int.Parse((new string(s)));
            }
        }
        return num;
    }
}
// @lc code=end

